[RFC] use alternatives to manage initramfs regeneration

ahesford opened the pull request:
Yes, I know that XBPS alternatives have some problems, but this seems to be the most sensible approach on balance. Rather than dealing with a bunch of potentially competing initramfs hooks, we can let all packages install custom hooks in `/usr/libexec` and use the alternatives system to symlink the preferred hooks in `/etc/kernel.d`. At that point, we can use the generic alternative script in the XBPS regenerator trigger and avoid the need for a conf file, although it falls back to legacy behavior if the alternative isn't set.

Another option I considered, but decided against, was creating a centralized `initramfs-kernel-hooks` package that installs a common hook and reads the `/etc/default/initramfs-regenerate` file used in the existing XBPS trigger to determine what to execute. All initramfs generators (currently only dracut and mkinitcpio) would have to depend on the common package. This has the drawback that all of the generation knowledge must be centralized instead of located in the package that adds the functionality.

This is not really tested at this point, but I want to get some early comments.

CameronNemo commented:
>This has the drawback that all of the generation knowledge must be centralized instead of located in the package that adds the functionality.

Is this much of a drawback? How many initramfs generators are we planning to add? More than a dozen? I can only count 5 proposed so far:

* tinyramfs
* mkinitcpio
* initramfs-tools
* booster
* dracut

CameronNemo commented on srcpkgs/dracut/template:
```diff
@@ -58,10 +62,14 @@ dracut-uefi_package() {
 	conf_files="/etc/default/dracut-uefi-hook"
+	alternatives="
+	 initramfs-hooks:/usr/libexec/dracut/kernel-uefi-post-install:/etc/kernel.d/post-install/20-initramfs
+	 initramfs-hooks:/usr/libexec/dracut/kernel-uefi-post-remove:/etc/kernel.d/post-remove/20-initramfs
```
Are UEFI generators really an "alternative" to traditional initramfs generators? Do they conflict? And what would happen in this case, if someone runs `xbps-install dracut-uefi`? They would have to switch the alternative after, right?

ahesford replied:
Yes and no. I was torn about adding an alternative here because it doesn't conflict in the same way as mkinitcpio/dracut, but presumably somebody interested in spitting out a UEFI bundle doesn't care about a separate initramfs in `/boot`. We could skip the alternative here and just install the hooks as they are now, or a user who wishes for both a separate initramfs and a UEFI bundle could manually link to the new hooks in `/usr/libexec`.

Motivation for this comes from https://github.com/void-linux/void-packages/pull/39282.

ahesford commented:
You're right that the number of initramfs generators that would be added is not likely to be large, so centralized logic wouldn't be a major bottleneck. That's why I mentioned this additional option: to try to find a consensus on the best way to make initramfs generators pluggable.

dmarto commented on srcpkgs/xbps-triggers/files/initramfs-regenerate:
```diff
@@ -37,36 +37,21 @@ case "$ACTION" in
+HOOK="etc/kernel.d/post-install/20-initramfs"
+if [ ! -x "${HOOK}" ]; then
+	# Otherwise, use legacy trigger behavior to find a preferred hook
+	[ -f etc/default/initramfs-regenerate ] && etc/default/initramfs-regenerate
```
missing `.`/`source`

ahesford replied:
Good catch, thanks

classabbyamp commented:
I think this is a reasonable approach

ahesford commented:
Did some testing and think this is ready for review. Fixed some bugs and dropped the alternative in `dracut-uefi`.

Looking for comments from @void-linux/pkg-committers.

thypon commented:
+1 on my side. I pushed for a similar change in the early days, but it was a bit aversed then.

classabbyamp commented:
I believe dracut and mkinitcpio need `xbps-triggers>=0.124` in depends now

ahesford commented:
An `xbps-triggers` dependency isn't needed because the existing trigger is independent of the initramfs kernel hooks. It will continue to work even if somebody updates only dracut or mkinitcpio.

classabbyamp commented:
tested and works for me, +1
